### PR TITLE
Multiple fixes to logic and layout

### PR DESCRIPTION
**Pull request by tenyo**
- Fixed subnet group generation - requires passing a list of subnets with the create request
- Added support for EngineVersion (defaults to latest)
- Added support for specifying security groups in create request
- Removed delete request body, and added `snapshot` query parameter
- Updated list/get logic and responses

Still need to implement rollback during create orchestration, if instance creation fails.

### CODE REVIEW

**Review by FishNecks — APPROVED** [review]
LGTM 👍

File: api/orchestration.go [hunk 1]
```diff
@@ -2,339 +2,192 @@ package api
...
-	// 		DBSubnetGroupCreateStatus, err := o.client.CreateDBSubnetGroup(ctx, &inputDBSubnetGroup)
+	req.Tags = normalizeTags(o.org, req.Tags)
 
-	if !ok {
-		return nil, apierror.New(apierror.ErrNotFound, "account doesnt exist", nil)
+	// check if a DBSubnetGroup exists, and create it if needed
+	dbSubnetGroupFound, err := o.dbSubnetGroupExists(ctx, dbSubnetGroupName(o.org))
+	if err != nil {
+		return nil, err
+	} else if !dbSubnetGroupFound {
```
Comment (fishnix) [inline api/orchestration.go]: this doesn't need to be an `else if`

Comment (fishnix) [reply api/orchestration.go]: you could also move that `log.Info` to the `dbSubnetGroupExists()` and make it a `debug` level to get rid of the `else`

File: api/orchestration.go [hunk 2]
```diff
@@ -2,339 +2,192 @@ package api
...
-			// failed to create db cluster (DBSubnetGroupNotFoundFault: DB subnet group
-			//'<search-name>' does not exist.
-			// status code: 404, request id: some request id)
-			regex := regexp.MustCompile("404")
-			match := regex.Match([]byte(fmt.Sprintln(err)))
-			if match == true {
-				log.Infof("DBSubnetGroup %s does not exist\n", aws.StringValue(search.DBSubnetGroupName))
-				log.Debugf("DBSubnetGroup %s does not exist: %v\n", aws.StringValue(search.DBSubnetGroupName), err)
-				return false
+		if aerr, ok := err.(awserr.Error); ok {
```
Comment (fishnix) [inline api/orchestration.go:94]: this should be done in the wrapper - `errors.go` - I see it doesn't exist currently.  It can be another PR.

File: api/orchestration.go [hunk 3]
```diff
@@ -2,339 +2,192 @@ package api
...
-			DBSubnetGroup:       *clusterCreateStatus.DBSubnetGroup,
-			DBInstances:         allInstances,
-		},
+	if snapshot {
+		input.SkipFinalSnapshot = aws.Bool(false)
+		input.FinalDBSnapshotIdentifier = aws.String("final-" + name)
 	}
 
-	marshaledData, err := json.Marshal(createOut)
+	_, err = o.client.DeleteDBCluster(ctx, &input)
```
Comment (fishnix) [inline api/orchestration.go]:
```golang
if _, err := .....
```

Comment (fishnix) [reply api/orchestration.go]: it is reasonable to return the deleted thing, im not sure if that's what was here. "ok" is fine too

File: api/orchestration.go (hunk 2 above)
Comment (fishnix) [reply api/orchestration.go:94]: there's a bunch of them, but ill just leave this one comment

File: api/types.go [hunk 4]
```diff
@@ -1,89 +1,76 @@
...
-	DBSubnetGroupARN         string
-	DBSubnetGroupDescription string
-	DBSubnetGroupName        string
-	SubnetGroupStatus        string
-	Subnets                  []*Subnet
-	VpcID                    string
-}
+// normalizeTags strips the org from the given tags and ensures it is set to the API org
+func normalizeTags(org string, tags []*Tag) []*Tag {
+	normalizedTags := []*Tag{}
```
Comment (fishnix) [inline api/types.go]: better to start with the org in the list, appends are "slow"

```golang
normalizedTags := []*Tag{
    {
        Key:   aws.String("spinup:org"),
        Value: aws.String(org),
    },
}
```

File: docdb/docdb.go [hunk 5]
```diff
@@ -62,30 +62,22 @@ func WithDefaultKMSKeyId(keyId string) DocDBOption {
 }
 
 // GetDBSubnetGroup gets documentDB DBSubnetGroup by name
-func (d *DocDB) GetDBSubnetGroup(ctx context.Context, input *docdb.DescribeDBSubnetGroupsInput) ([]*docdb.DBSubnetGroup, error) {
-	if input == nil {
-		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", nil)
-	}
-
-	log.Infof("searching for documentDB dbsubnetgroup: %s\n", aws.StringValue(input.DBSubnetGroupName))
+func (d *DocDB) GetDBSubnetGroup(ctx context.Context, name string) ([]*docdb.DBSubnetGroup, error) {
```
Comment (fishnix) [inline docdb/docdb.go:65]: 🎉

File: docdb/docdb.go [hunk 6]
```diff
@@ -62,30 +62,22 @@ func WithDefaultKMSKeyId(keyId string) DocDBOption {
...
 
-// ListDB lists documentdb clusters
-func (d *DocDB) ListDB(ctx context.Context, input *docdb.DescribeDBClustersInput) (*docdb.DescribeDBClustersOutput, error) {
-	if input == nil {
-		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", nil)
-	}
-
-	log.Infoln("listing documentDB clusters and instance(s)")
+// ListDocDBs lists documentDB clusters
+func (d *DocDB) ListDocDBClusters(ctx context.Context) ([]string, error) {
```
Comment (fishnix) [inline docdb/docdb.go:79]: 🎉

File: session/session.go [hunk 7]
```diff
@@ -42,14 +42,12 @@ func New(opts ...SessionOption) Session {
 
 func WithCredentials(key, secret, token string) SessionOption {
 	return func(s *Session) {
-		log.Debugf("setting credentials with key id %s", key)
```
Comment (fishnix) [inline session/session.go]: why remove these debug logs?

File: sts/sts.go [hunk 8]
```diff
@@ -59,14 +59,10 @@ func (s *STS) AssumeRole(ctx context.Context, input *sts.AssumeRoleInput) (*sts.
 
 	log.Infof("assuming role '%s' with session name '%s'", aws.StringValue(input.RoleArn), aws.StringValue(input.RoleSessionName))
 
-	log.Debugf("assuming role %s with input %+v", aws.StringValue(input.RoleArn), input)
```
Comment (fishnix) [inline sts/sts.go]: same

**Review by fishnix — COMMENTED** [review]
lookingood

File: session/session.go (hunk 7 above)
Comment (tenyo) [reply session/session.go]: haha, i got carried away 😄 i'll add them back

File: api/orchestration.go (hunk 2 above)
Comment (tenyo) [reply api/orchestration.go:94]: yup, i realized we're missing the DocDB errors, i'll add them separately

File: api/orchestration.go (hunk 1 above)
Comment (tenyo) [reply api/orchestration.go]: We need the `else if` to check if the subnet group exists.
I wanted to explicitly log it as Info since it's useful to see which subnet group we are using exactly, especially in the beginning as we start using this

Comment (fishnix) [reply api/orchestration.go]: > We need the else if to check if the subnet group exists.

unless I'm being dense (possible),  i don't think you do:

```golang
	// check if a DBSubnetGroup exists, and create it if needed
	dbSubnetGroupFound, err := o.dbSubnetGroupExists(ctx, dbSubnetGroupName(o.org))
	if err != nil {
		return nil, err
	}

        if !dbSubnetGroupFound {
		if err := o.dbSubnetGroupCreate(ctx, dbSubnetGroupName(o.org), req.SubnetIds); err != nil {
			return nil, err
		}
	} else {
		log.Infof("subnet group %s already exists, will use it for this docdb cluster", dbSubnetGroupName(o.org))
	}
```

> I wanted to explicitly log it as Info since it's useful to see which subnet group we are using exactly, especially in the beginning as we start using this

I thought you might be able to put the log message in `dbSubnetGroupExists()` instead since it would avoid the else, but this is ok too 👍

Comment (fishnix) [reply api/orchestration.go]: maybe you were thinking to do this?

```golang
	// check if a DBSubnetGroup exists, and create it if needed
	if dbSubnetGroupFound, err := o.dbSubnetGroupExists(ctx, dbSubnetGroupName(o.org)); err != nil {
		return nil, err
	} else if !dbSubnetGroupFound {
		if err := o.dbSubnetGroupCreate(ctx, dbSubnetGroupName(o.org), req.SubnetIds); err != nil {
			return nil, err
		}
	} else {
		log.Infof("subnet group %s already exists, will use it for this docdb cluster", dbSubnetGroupName(o.org))
	}
```
